services/management/grpc/annotation_server.go

```diff
@@ -54,9 +58,88 @@ func (as *AnnotationServer) AddAnnotation(ctx context.Context, req *managementpb
 		return nil, status.Error(codes.Unauthenticated, "Authorization error.")
 	}
 
-	_, err := as.grafanaClient.CreateAnnotation(ctx, req.Tags, time.Now(), req.Text, authorizationHeaders[0])
-	if err != nil {
-		return nil, err
+	var service, node bool
+	var serviceErr, nodeErr error
+	if len(req.ServiceName) > 0 {
+		service = true
+	}
+	if req.NodeName != "" {
+		node = true
+	}
+
+	if service {
+		serviceErr = findService(as.db, req.ServiceName)
+	}
+
+	if node {
+		nodeErr = findNode(as.db, req.NodeName)
+	}
+
+	if serviceErr != nil && nodeErr != nil || serviceErr != nil {
+		return nil, serviceErr
+	}
+
+	if nodeErr != nil {
+		return nil, nodeErr
+	}
+
+	if service {
+		for _, sn := range req.ServiceName {
+			text := fmt.Sprintf("%s (Service Name: %s)", req.Text, sn)
+			_, err := as.grafanaClient.CreateAnnotation(ctx, req.Tags, time.Now(), text, authorizationHeaders[0])
+			if err != nil {
```

Comment:
You need to append to user-defined tags service names and node names

curl 'http://admin:admin@localhost/graph/api/annotations' -H 'Content-Type: application/json;charset=UTF-8' --data-binary '{"time":1585053244718,"tags":["user specified tag1", "user specified tag2", "ps1", "ps2", "ps3", "node1"],"text":"Alarm!!! (Service Name: ps1,ps2,ps3, Node Name:node1)"}'